### ISSUE TITLE
Shell scripts for building / running server in docker fail

### ISSUE THREAD

**Issue by chvanlennep** [rcpch/digital-growth-charts-server#131]
**Environment:**

> System:
>     OS: macOS 11.1
>     CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
>     Memory: 220.33 MB / 16.00 GB
>     Shell: 5.8 - /bin/zsh
>   Binaries:
>     Node: 14.15.3 - ~/.nvm/versions/node/v14.15.3/bin/node
>     npm: 6.14.9 - ~/.nvm/versions/node/v14.15.3/bin/npm
>     Watchman: 4.9.0 - /usr/local/bin/watchman
>   Managers:
>     CocoaPods: 1.10.0 - /usr/local/bin/pod
>     Homebrew: 3.0.1 - /usr/local/bin/brew
>     pip3: 21.0.1 - /usr/local/bin/pip3
>     RubyGems: 3.0.3 - /usr/bin/gem
>   Utilities:
>     Make: 3.81 - /usr/bin/make
>     GCC: 4.2.1 - /usr/bin/gcc
>     Git: 2.24.3 - /usr/bin/git
>     Clang: 1200.0.32.28 - /usr/bin/clang
>   Servers:
>     Apache: 2.4.46 - /usr/sbin/apachectl
>   Virtualization:
>     Docker: 20.10.2 - /usr/local/bin/docker
>   SDKs:
>     iOS SDK:
>       Platforms: iOS 14.3, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
>   IDEs:
>     Android Studio: 4.1 AI-201.8743.12.41.7042882
>     Nano: 2.0.6 - /usr/bin/nano
>     VSCode: 1.53.0 - /usr/local/bin/code
>     Vim: 8.2 - /usr/bin/vim
>     Xcode: 12.3/12C33 - /usr/bin/xcodebuild
>   Languages:
>     Bash: 3.2.57 - /bin/bash
>     Java: 15.0.1 - /usr/bin/javac
>     Perl: 5.28.2 - /usr/bin/perl
>     PHP: 7.3.24 - /usr/bin/php
>     Python: 2.7.16 - /usr/bin/python
>     Python3: 3.9.1 - /usr/local/bin/python3
>     Ruby: 2.6.3 - /usr/bin/ruby
>   Databases:
>     SQLite: 3.32.3 - /usr/bin/sqlite3
>   Browsers:
>     Chrome: 88.0.4324.150
>     Edge: 88.0.705.68
>     Firefox: 84.0.2
>     Safari: 14.0.2

**Running rebuild-docker script fails. Output:**

> Terminal: 
> 
>  > [internal] load build definition from Dockerfile:
> ------
> failed to solve with frontend dockerfile.v0: failed to read dockerfile: error from sender: open .Trash: operation not permitted
> Saving session...
> ...copying shared history...
> ...saving history...truncating history files...
> ...completed.
> 
> [Process completed]

I think this might be because of the default directory the script runs from? I added in some code to change up a directory and the script ran without issue:

```
#!/bin/bash

app="rcpch-dgc-server"

cd $(dirname $0)
cd ../

if [ "$(docker images -q ${app} 2>/dev/null)" != "" ]; then
    docker container stop ${app} &&
        docker container rm ${app} &&
        docker image rm ${app}
fi

docker build -t ${app} .
```

**Running docker / server**

After successfully building docker image, start-docker scripts completes without errors, but no working flask server. When attempting to run start-server:

flask: command not found

I was able to get docker container running and server running with the following script:

```
#!/bin/bash

app="rcpch-dgc-server"

docker run -it -p 5000:5000 --name ${app} ${app} 
```

**Comment by pacharanero** [comment]
>When attempting to run start-server `flask: command not found`

Have you run `pip install -r requirements.txt`? This step installs all the necessary python packages.
Updated dev docs are coming but you can preview them here https://github.com/rcpch/digital-growth-charts-server/blob/term-reference-removed/docs/README.md#setup-for-this-repository

**Comment by chvanlennep** [comment]
Docker already does this when the container is created- it's included in the DockerFile config.

 I also tried running the start-server script from within the container, but that did not work either.

**Comment by pacharanero** [comment]
I think you are mixing up the purposes of the scripts.

`s/start-server` is only for when you are running the server _natively_ on your machine. Don't use it when you are using the Docker container.

`s/rebuild-docker` and `s/start-docker` are for when you are using the Docker container. They automatically should start Flask inside the Docker container.

I will have a look at the docs and check it's clearly written

**Comment by chvanlennep** [comment]
Thanks Marcus- I appreciate that the start-server script is designed for native environment installs.

I tried that script only because of the problems I got when using the start-docker script, thinking that perhaps the server needed to be started once the container was running.

Here are the error headings in the html when navigating to localhost:5000 after using the current start-docker script:

> flask.cli.NoAppException
> flask.cli.NoAppException: Could not import "app".

As I said above, I got errors with the docker scripts as written but could get them to work with the modifications as written.

I've done some more digging around and I think it's related to opening the scripts from a GUI (as I'm using in macOS) and not the command line.

**Comment by pacharanero** [comment]
Ah OK thanks. I will look at your modifications and see if they work on my machine.

**Comment by AcidUK** [comment]
The Dockerfile copies the app into the /app folder during the build phase. The start-docker script is mapping the current directory over the /app folder, which essentially is masking the original /app folder. 

This has the benefit of updating the code running within docker if you edit your local files, however Is like the cause of the issue.

I suspect in your development environment @pacharanero you already have the requirements installed, however @chvanlennep does not, and so the container fails.

Since the role of the docker container in this situation is provide a homogenous runtime environment it would probably be best to remove the volume mapping at https://github.com/rcpch/digital-growth-charts-server/blob/86eb3f89b62bb0c27a15720ba5fe8ae9762b16e0/s/start-docker#L14 (this is essentially what @chvanlennep has done)

Happy to submit a PR if helpful.

On a side note, is there documentation about how to use the the build scripts, or is it assumed that people will know (for example if you run them from ./s/ they will fail due to the Dockerfile being in the parent directory)? Again I'm happy to add this documentation via PR if you'd like, and/or split into a new issue.

**Comment by pacharanero** [comment]
Hi @AcidUK long time no see!
You make good points
Very happy to receive a PR that fixes it. Additions to the documentation would also be welcome.
We have just moved over to using Material for MkDocs and the doco is here:
https://github.com/rcpch/digital-growth-charts-documentation/blob/alpha/docs/developer/api-docker.md

**Comment by pacharanero** [comment]
closed by #161
